apps/jupyter/extensions/notebook-runner/src/notebook-runner-state.ts

```diff
@@ -0,0 +1,276 @@
+
+  private _resolveOtterSessionContext: (value: SessionContext) => void =
+    () => {};
+
+  public otterSessionContext: Promise<SessionContext> =
+    new Promise<SessionContext>((resolve) => {
+      this._resolveOtterSessionContext = resolve;
+    });
+
+  private commListeners: CommListener[] = [];
+
+  constructor(
+    protected readonly app: JupyterFrontEnd,
+    protected readonly documentManager: IDocumentManager,
+    protected readonly notebookTracker: INotebookTracker,
+    public allowNextNotebookInParallel: boolean = false,
+  ) {
+    this.init();
+  }
+
+  public addCommListener(listener: CommListener): void {
+    this.commListeners.push(listener);
+  }
+
+  public removeCommListener(listener: CommListener): void {
+    this.commListeners = this.commListeners.filter((l) => l !== listener);
+  }
+
+  private async init(): Promise<void> {
+    // eslint-disable-next-line no-async-promise-executor
+    const initPromise = new Promise<SessionContext>(async (resolve, reject) => {
+      let sessionContext: SessionContext | undefined = undefined;
+
+      try {
+        const serviceManager = this.app.serviceManager;
+
+        console.log("Initializing Otter session context..");
+        sessionContext = new SessionContext({
+          sessionManager: serviceManager.sessions,
+          specsManager: serviceManager.kernelspecs,
+          name: "otter-session",
+          type: "notebook",
+          kernelPreference: {
+            shouldStart: true,
+            canStart: true,
+            autoStartDefault: true,
+          },
+        });
+
+        // Initialize the session
+        console.debug("Initializing otter context..");
+        await sessionContext.initialize();
+
+        console.log("Adding kernel listeners to otter context..");
+
+        await addKernelListeners(sessionContext, async (kernel) => {
+          console.debug(
+            "Kernel is ready:",
+            kernel.name,
+            "attaching listeners..",
+          );
+          this.registerCommTarget(kernel);
+          await this.prepareOtterKernel(kernel);
+        });
+
+        console.debug("Otter context initialized:", sessionContext);
+        resolve(sessionContext);
+        return;
+      } catch (e) {
+        if (sessionContext) {
+          let disposed: () => void = () => {};
+          const isDisposed = new Promise<void>((resolve) => {
+            disposed = resolve;
+          });
+          sessionContext.disposed.connect(disposed);
+          sessionContext?.dispose();
+
+          await isDisposed;
+        }
+
+        reject(e);
+        return;
+      }
+    });
+
+    try {
+      const sessionContext = await initPromise;
+      this._resolveOtterSessionContext(sessionContext);
+    } catch (e) {
+      console.error("Error initializing otter context:", e, ". Restarting..");
+      return this.init();
+    }
+  }
+
+  private registerCommTarget(kernel: IKernelConnection): void {
```

Comment:
when reading this code as a newbie, I have no idea what a `CommTarget` is or what this does. The naming is coherent with the kernel API, but we may want to use a combination of naming and/or documentation to make the intention here more explicit?